[Merged by Bors] - feat(MeasureTheory): TendstoInMeasure gives ae unique limits; TendstoInMeasure equivalent to subsequences converging ae

pfaffelh opened the pull request:
feat: change mostly in MeasureTheory.Function.TendstoInMeasure
Two new features:
1) Limits in Measure are ae unique
2) Convergence in Measure is equivalent to the proof that every subsequence has a further subsequence which converges (back-direction is new)


---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

JasonKYi commented:
Can you merge master and resolve the merge conflict?

pfaffelh commented:
> Can you merge master and resolve the merge conflict?

I would really like to do that. However, I cannot see the merge conflict in my local vscode, so I need to find out what the conflict is...

grunweg commented:
> > Can you merge master and resolve the merge conflict?
> 
> I would really like to do that. However, I cannot see the merge conflict in my local vscode, so I need to find out what the conflict is...

I've taken the liberty to just merge master - let me know if that was premature and I can revert this!

I think the conflict was spurious, in the sense that git's default algorithm could solve this on its own. (Hence, I would not get any conflicts in VS Code.) In this case, I think you can click on the "merge" button directly. (Or, if you work from the terminal, typing `git merge master` will already succeed, you can confirm the commit message and push.)

pfaffelh commented:
> > > Can you merge master and resolve the merge conflict?
> > 
> > 
> > I would really like to do that. However, I cannot see the merge conflict in my local vscode, so I need to find out what the conflict is...
> 
> I've taken the liberty to just merge master - let me know if that was premature and I can revert this!
> 
> I think the conflict was spurious, in the sense that git's default algorithm could solve this on its own. (Hence, I would not get any conflicts in VS Code.) In this case, I think you can click on the "merge" button directly. (Or, if you work from the terminal, typing `git merge master` will already succeed, you can confirm the commit message and push.)

Many thanks, that helped a lot!

JasonKYi commented on Mathlib/Data/ENNReal/Basic.lean:
```diff
@@ -208,6 +208,14 @@ theorem coe_toNNReal : ∀ {a : ℝ≥0∞}, a ≠ ∞ → ↑a.toNNReal = a
 
+@[simp]
+theorem coe_of_fun_toNNReal {ι : Type*} {f : ι → ℝ≥0∞} (hf : ∀ x, f x ≠ ⊤) :
+    (fun (x : NNReal) => (x : ENNReal)) ∘ ENNReal.toNNReal ∘ f = f := by
+  ext x
+  change (ENNReal.toNNReal (f x) : ℝ≥0∞) = f x
+  rw [coe_toNNReal]
+  exact hf x
```
```suggestion
    (fun (x : ℝ≥0) => (x : ℝ≥0∞)) ∘ ENNReal.toNNReal ∘ f = f := by
  ext x
  simp [coe_toNNReal (hf x)]
```

JasonKYi commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -62,6 +66,50 @@ theorem tendstoInMeasure_iff_norm [SeminormedAddCommGroup E] {l : Filter ι} {f
+    {x : MeasurableSpace α} {μ : Measure α} {g h : α → E} {f : ι → α → E} {l : Filter ι} [l.NeBot]
+    (hg : ∀ᵐ ω ∂μ, Filter.Tendsto (fun i => f i ω) l (nhds (g ω))) (hh :
+    ∀ᵐ ω ∂μ, Filter.Tendsto (fun i => f i ω) l (nhds (h ω))) : g =ᵐ[μ] h := by
+  filter_upwards [hg, hh] with ω hg1 hh1
+  exact tendsto_nhds_unique hg1 hh1
```
```suggestion
  filter_upwards [hg, hh] with ω hg1 hh1 using tendsto_nhds_unique hg1 hh1
```
Also the spelling for ae convergence is `tendsto_ae` so I would change the name of this lemma to `tendsto_ae_unique`

sgouezel commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -62,6 +66,50 @@ theorem tendstoInMeasure_iff_norm [SeminormedAddCommGroup E] {l : Filter ι} {f
+-- The ae-limit is ae-unique.
+theorem ae_unique_of_ae_limit {α ι E : Type*} [TopologicalSpace E] [T2Space E]
+    {x : MeasurableSpace α} {μ : Measure α} {g h : α → E} {f : ι → α → E} {l : Filter ι} [l.NeBot]
+    (hg : ∀ᵐ ω ∂μ, Filter.Tendsto (fun i => f i ω) l (nhds (g ω))) (hh :
```
Could you use `𝓝` instead of `nhds` (obtained by typing \nhds) here and elsewhere? Also, your assumption `hh` is split among two lines, while it would fit on one line.

sgouezel commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -62,6 +66,50 @@ theorem tendstoInMeasure_iff_norm [SeminormedAddCommGroup E] {l : Filter ι} {f
+
+theorem TendstoInMeasure_of_FiniteMeasure [Dist E] {_ : MeasurableSpace α} {μ : Measure α}
+    [hfin: MeasureTheory.IsFiniteMeasure μ] {f : ι → α → E} {l : Filter ι} {g : α → E} :
+TendstoInMeasure μ f l g ↔ TendstoInMeasure' μ f l g := by
```
```suggestion
    TendstoInMeasure μ f l g ↔ TendstoInMeasure' μ f l g := by
```
Intendation in theorem statements should be 4 spaces.

sgouezel commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -62,6 +66,50 @@ theorem tendstoInMeasure_iff_norm [SeminormedAddCommGroup E] {l : Filter ι} {f
+    exact h ε hε
+
+-- Convergence in measure is stable under taking subsequences.
+lemma sub_TendstoInMeasure' [Dist E] {f : ι → α → E}  {g : α → E} {u v: Filter ι} (huv : v ≤ u)
```
```suggestion
lemma sub_TendstoInMeasure' [Dist E] {f : ι → α → E}  {g : α → E} {u v : Filter ι} (huv : v ≤ u)
```

JasonKYi commented:
Did you make changes locally? I think you might have forgotten to push.

pfaffelh commented:
> Did you make changes locally? I think you might have forgotten to push.

Yes, indeed. Now it is pushed...

pfaffelh commented:
> Is this ready to be reviewed again? If so, please remove the awaiting-author label and readd the awaiting-review label.

Thanks, I am still new to making PRs...

JasonKYi commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -62,6 +66,49 @@ theorem tendstoInMeasure_iff_norm [SeminormedAddCommGroup E] {l : Filter ι} {f
 #align measure_theory.tendsto_in_measure_iff_norm MeasureTheory.tendstoInMeasure_iff_norm
 
+-- The ae-limit is ae-unique.
+theorem tendsto_ae_unique {α ι E : Type*} [TopologicalSpace E] [T2Space E]
```
Can you see if you can move this to either where `tendsto_nhds_unique` is or where the `ae` filter is defined?

pfaffelh replied:
I agree that it is/was not in a good place. I think `MeasureTheory.Measure.Function.AEEqFun.lean` is a good place and I will put it there.

JasonKYi commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -62,6 +66,49 @@ theorem tendstoInMeasure_iff_norm [SeminormedAddCommGroup E] {l : Filter ι} {f
+
+/-- This notion is helpful for finite measures since we don't have to deal with the
+possibility that some set measures to ∞ -/
+def TendstoInMeasure' [Dist E] {_ : MeasurableSpace α} (μ : Measure α) (f : ι → α → E)
```
I'm not convinced that this definition is useful. Can you explain a bit more why it is needed?

pfaffelh replied:
It boils down to having a coercion `NNRead.toReal` but not a coercion `ENNReal.toReal`. Since the former exists, it is possible to compute stuff more easily when using `TendstoInMeasure'` rather than `TendstoInMeasure`. Of course there are other ways to avoid having this definition, but ir really helped me computing things faster.

JasonKYi replied:
I think rather than making a new definition, it is better to replace the right hand side of `TendstoInMeasure_of_FiniteMeasure` by ` ∀ ε, 0 < ε → Tendsto (ENNReal.toNNReal ∘ (fun i => (μ { x | ε ≤ dist (f i x) (g x) }))) l (𝓝 0)`

pfaffelh replied:
This is also a good idea. I not deleted the definition and followed the suggestion.

JasonKYi commented on Mathlib/Data/ENNReal/Basic.lean:
```diff
@@ -194,6 +194,12 @@ theorem coe_toNNReal : ∀ {a : ℝ≥0∞}, a ≠ ∞ → ↑a.toNNReal = a
   | ⊤, h => (h rfl).elim
 
+@[simp]
+theorem coe_of_fun_toNNReal {ι : Type*} {f : ι → ℝ≥0∞} (hf : ∀ x, f x ≠ ⊤) :
```
```suggestion
theorem coe_of_fun_toNNReal {ι : Type*} {f : ι → ℝ≥0∞} (hf : ∀ x, f x ≠ ∞) :
```

pfaffelh replied:
Thanks, changed.

JasonKYi commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -32,6 +33,9 @@ convergence in measure and other notions of convergence.
   which converges in measure to `g`, then `f` has a subsequence which convergence almost
   everywhere to `g`.
+* `MeasureTheory.exists_seq_tendstoInMeasure_atTop_iff`: for a sequence of functions `f`,
+convergence in Measure is equivalent to the fact that every subsequence has another subsequence
```
```suggestion
convergence in measure is equivalent to the fact that every subsequence has another subsequence
```

pfaffelh replied:
Changed.

JasonKYi commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -226,12 +289,125 @@ theorem TendstoInMeasure.exists_seq_tendstoInMeasure_atTop {u : Filter ι} [NeBo
+
+/- TendstoInMeasure is equivalent to a proof  that every subsequence has another subsequence
+which converges almost surely. -/
+theorem exists_seq_tendstoInMeasure_atTop_iff     (hfin : MeasureTheory.IsFiniteMeasure μ)
```
```suggestion
theorem exists_seq_tendstoInMeasure_atTop_iff (hfin : MeasureTheory.IsFiniteMeasure μ)
```

pfaffelh replied:
Changed.

JasonKYi commented on Mathlib/Topology/Instances/ENNReal.lean:
```diff
@@ -106,6 +111,21 @@ theorem tendsto_toNNReal {a : ℝ≥0∞} (ha : a ≠ ∞) :
   exact tendsto_id
 
+theorem tendsto_toNNReal_iff {ι : Type*} {f : ι → ℝ≥0∞} {u : Filter ι} {a : ℝ≥0∞}
+    (ha : a ≠ ⊤) (hf : ∀ x, f x ≠ ⊤) : Tendsto f u (nhds a) ↔
```
```suggestion
    (ha : a ≠ ∞) (hf : ∀ x, f x ≠ ∞) : Tendsto f u (nhds a) ↔
```

JasonKYi commented on Mathlib/Topology/Instances/ENNReal.lean:
```diff
@@ -106,6 +111,21 @@ theorem tendsto_toNNReal {a : ℝ≥0∞} (ha : a ≠ ∞) :
+    exact h2
+
+theorem tendsto_toNNReal_iff' {ι : Type*} {f : ι → ℝ≥0∞} {u : Filter ι} {a : ℝ≥0}
+    (hf : ∀ x, f x ≠ ⊤): Tendsto f u (nhds a) ↔ Tendsto (ENNReal.toNNReal ∘ f ) u (nhds a) := by
```
```suggestion
    (hf : ∀ x, f x ≠ ∞): Tendsto f u (𝓝 a) ↔ Tendsto (ENNReal.toNNReal ∘ f ) u (𝓝 a) := by
```

JasonKYi commented on Mathlib/Topology/MetricSpace/Pseudo/Defs.lean:
```diff
@@ -822,6 +822,15 @@ theorem tendsto_atTop' [Nonempty β] [SemilatticeSup β] [NoMaxOrder β] {u : β
+/-- This variant can be useful for proofs by contradiction.
+-/
+lemma false_of_Tendsto_of_boundBelow  [Nonempty β] [SemilatticeSup β] {u : β → α}
+    {a : α} {δ : ℝ} (hδ: (0 : ℝ) < δ) (hu1 : Tendsto u atTop (nhds a))
```
```suggestion
    {a : α} {δ : ℝ} (hδ: (0 : ℝ) < δ) (hu1 : Tendsto u atTop (𝓝 a))
```

pfaffelh commented on Mathlib/Topology/Instances/ENNReal.lean:
```diff
@@ -74,6 +74,11 @@ theorem tendsto_coe {f : Filter α} {m : α → ℝ≥0} {a : ℝ≥0} :
 theorem continuous_coe : Continuous ((↑) : ℝ≥0 → ℝ≥0∞) :=
   embedding_coe.continuous
 
+lemma tendsto_coe_id {a : ℝ≥0∞} (ha : a ≠ ⊤) : Tendsto (fun (x : NNReal) => (x : ENNReal))
```
Thanks, changed.

JasonKYi commented on Mathlib/MeasureTheory/Function/AEEqFun.lean:
```diff
@@ -863,6 +863,17 @@ theorem coeFn_posPart (f : α →ₘ[μ] γ) : ⇑(posPart f) =ᵐ[μ] fun a =>
+-- The ae-limit is ae-unique.
+theorem tendsto_ae_unique {α ι E : Type*} [TopologicalSpace E] [T2Space E]
+    {x : MeasurableSpace α} {μ : Measure α} {g h : α → E} {f : ι → α → E} {l : Filter ι} [l.NeBot]
+    (hg : ∀ᵐ ω ∂μ, Filter.Tendsto (fun i => f i ω) l (𝓝 (g ω)))
+    (hh : ∀ᵐ ω ∂μ, Filter.Tendsto (fun i => f i ω) l (𝓝 (h ω))) : g =ᵐ[μ] h := by
```
```suggestion
    (hg : ∀ᵐ ω ∂μ, Tendsto (fun i => f i ω) l (𝓝 (g ω)))
    (hh : ∀ᵐ ω ∂μ, Tendsto (fun i => f i ω) l (𝓝 (h ω))) : g =ᵐ[μ] h := by
```
Does this work? I think `Filter` has already been opened here.

pfaffelh replied:
Yes, it does!

JasonKYi commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -58,6 +62,37 @@ theorem tendstoInMeasure_iff_norm [SeminormedAddCommGroup E] {l : Filter ι} {f
 
+theorem TendstoInMeasure_of_FiniteMeasure [Dist E] {_ : MeasurableSpace α} {μ : Measure α}
+    [hfin: MeasureTheory.IsFiniteMeasure μ] {f : ι → α → E} {l : Filter ι} {g : α → E} :
+    TendstoInMeasure μ f l g ↔ ∀ ε, 0 < ε → Tendsto (ENNReal.toNNReal ∘ (fun i =>
+    (μ { x | ε ≤ dist (f i x) (g x) }))) l (𝓝 0) := by
```
```suggestion
    TendstoInMeasure μ f l g ↔ ∀ ε, 0 < ε → Tendsto (fun i =>
    (μ { x | ε ≤ dist (f i x) (g x) }).toNNReal) l (𝓝 0) := by
```
This would be a more natural way of formulating

pfaffelh replied:
Ok, changed.

JasonKYi commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -58,6 +62,37 @@ theorem tendstoInMeasure_iff_norm [SeminormedAddCommGroup E] {l : Filter ι} {f
+-- Convergence in measure is stable under taking subsequences.
+lemma sub_TendstoInMeasure' [Dist E] {f : ι → α → E}  {g : α → E} {u v : Filter ι} (huv : v ≤ u)
+    (hg :  TendstoInMeasure μ f u g) : TendstoInMeasure μ f v g :=
+  fun ε hε => Tendsto.mono_left (hg ε hε) huv
```
```suggestion
  fun ε hε => (hg ε hε).mono_left huv
```

pfaffelh replied:
Ok, changed.

JasonKYi commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -58,6 +62,37 @@ theorem tendstoInMeasure_iff_norm [SeminormedAddCommGroup E] {l : Filter ι} {f
       ∀ ε, 0 < ε → Tendsto (fun i => μ { x | ε ≤ ‖f i x - g x‖ }) l (𝓝 0) := by
   simp_rw [TendstoInMeasure, dist_eq_norm]
 
+theorem TendstoInMeasure_of_FiniteMeasure [Dist E] {_ : MeasurableSpace α} {μ : Measure α}
```
```suggestion
theorem tendstoInMeasure_iff_tendsto_toNNReal [Dist E] {_ : MeasurableSpace α} {μ : Measure α}
```

pfaffelh replied:
Ok, changed the name.

JasonKYi commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -58,6 +62,37 @@ theorem tendstoInMeasure_iff_norm [SeminormedAddCommGroup E] {l : Filter ι} {f
+    exact h ε hε
+
+-- Convergence in measure is stable under taking subsequences.
+lemma sub_TendstoInMeasure' [Dist E] {f : ι → α → E}  {g : α → E} {u v : Filter ι} (huv : v ≤ u)
```
```suggestion
lemma tendstoInMeasure_mono [Dist E] {f : ι → α → E}  {g : α → E} {u v : Filter ι} (huv : v ≤ u)
```

pfaffelh replied:
Ok, changed.

JasonKYi commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -58,6 +62,37 @@ theorem tendstoInMeasure_iff_norm [SeminormedAddCommGroup E] {l : Filter ι} {f
+    (hg :  TendstoInMeasure μ f u g) : TendstoInMeasure μ f v g :=
+  fun ε hε => Tendsto.mono_left (hg ε hε) huv
+
+lemma subseqTendsto_of_TendstoInMeasure [Dist E] {f : ℕ → α → E}  {g : α → E} {ns : ℕ → ℕ}
+    (hns : StrictMono ns) (hg :  TendstoInMeasure μ f atTop g) :
```
```suggestion
lemma TendstoInMeasure.subseq [Dist E] {f : ℕ → α → E}  {g : α → E} {ns : ℕ → ℕ}
    (hg :  TendstoInMeasure μ f atTop g) (hns : StrictMono ns) :
```

JasonKYi commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -58,6 +62,37 @@ theorem tendstoInMeasure_iff_norm [SeminormedAddCommGroup E] {l : Filter ι} {f
+    (hg :  TendstoInMeasure μ f u g) : TendstoInMeasure μ f v g :=
+  fun ε hε => Tendsto.mono_left (hg ε hε) huv
+
+lemma subseqTendsto_of_TendstoInMeasure [Dist E] {f : ℕ → α → E}  {g : α → E} {ns : ℕ → ℕ}
```
```suggestion
lemma TendstoInMeasure.subseq [Dist E] {f : ℕ → α → E}  {g : α → E} {ns : ℕ → ℕ}
```

pfaffelh replied:
Changed.

JasonKYi commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -58,6 +62,37 @@ theorem tendstoInMeasure_iff_norm [SeminormedAddCommGroup E] {l : Filter ι} {f
+  intro ε hε
+  apply Filter.Tendsto.comp (hg ε hε) (StrictMono.tendsto_atTop hns)
+
+lemma subseq_TendstoInMeasure' [Dist E] {f : ι → α → E}  {g : α → E} {u : Filter ι}
```
```suggestion
lemma TendstoInMeasure.subseq' [Dist E] {f : ι → α → E}  {g : α → E} {u : Filter ι}
```

pfaffelh replied:
Changed.

JasonKYi commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -88,6 +123,23 @@ theorem congr_right (h : g =ᵐ[μ] g') (h_tendsto : TendstoInMeasure μ f l g)
     TendstoInMeasure μ f l g' :=
   h_tendsto.congr (fun _ => EventuallyEq.rfl) h
 
+-- Convergence in measure is stable under taking subsequences.
+theorem sub_TendstoInMeasure' {f : ι → α → E}  {g : α → E} {u v: Filter ι} (huv : v ≤ u)
+    (hg :  TendstoInMeasure μ f u g) : TendstoInMeasure μ f v g :=
+  fun ε hε => Tendsto.mono_left (hg ε hε) huv
```
Isn't this lemma a duplicate of the above?

pfaffelh replied:
Yes, you are right. I deleted them. Thanks!

JasonKYi commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -218,6 +270,12 @@ theorem TendstoInMeasure.exists_seq_tendsto_ae (hfg : TendstoInMeasure μ f atTo
+  obtain ⟨ns, h_tendsto_ns⟩ : ∃ ns : ℕ → ι, Tendsto ns atTop u := exists_seq_tendsto u
+  exact ⟨ns, h_tendsto_ns, fun ε hε => (hfg ε hε).comp h_tendsto_ns⟩
+
+theorem TendstoInMeasure.exists_seq_tendstoInMeasure_atTop' {u : Filter ι} [NeBot u]
```
This is a duplicate of the above lemma

JasonKYi commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -218,6 +270,12 @@ theorem TendstoInMeasure.exists_seq_tendsto_ae (hfg : TendstoInMeasure μ f atTo
+  obtain ⟨ns, h_tendsto_ns⟩ : ∃ ns : ℕ → ι, Tendsto ns atTop u := exists_seq_tendsto u
+  exact ⟨ns, h_tendsto_ns, fun ε hε => (hfg ε hε).comp h_tendsto_ns⟩
+
+theorem TendstoInMeasure.exists_seq_tendstoInMeasure_atTop' {u : Filter ι} [NeBot u]
```
I think this lemma is strictly weaker than the previous. Can you remove it?

pfaffelh replied:
Yes, I removed it. The weaker version was part of the file before my PR.

JasonKYi commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -226,12 +284,125 @@ theorem TendstoInMeasure.exists_seq_tendstoInMeasure_atTop {u : Filter ι} [NeBo
 theorem TendstoInMeasure.exists_seq_tendsto_ae' {u : Filter ι} [NeBot u] [IsCountablyGenerated u]
     {f : ι → α → E} {g : α → E} (hfg : TendstoInMeasure μ f u g) :
     ∃ ns : ℕ → ι, ∀ᵐ x ∂μ, Tendsto (fun i => f (ns i) x) atTop (𝓝 (g x)) := by
```
```suggestion
    ∃ (ns : ℕ → ι) (_ : Tendsto ns atTop u), ∀ᵐ x ∂μ, Tendsto (fun i => f (ns i) x) atTop (𝓝 (g x)) := by
```
This should be free now

pfaffelh replied:
Yes, you are right, this now also works. Added...

JasonKYi commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -58,6 +62,41 @@ theorem tendstoInMeasure_iff_norm [SeminormedAddCommGroup E] {l : Filter ι} {f
       ∀ ε, 0 < ε → Tendsto (fun i => μ { x | ε ≤ ‖f i x - g x‖ }) l (𝓝 0) := by
   simp_rw [TendstoInMeasure, dist_eq_norm]
 
+theorem TendstoInMeasure_iff_tendsto_toNNReal [Dist E] {_ : MeasurableSpace α} {μ : Measure α}
```
More specifically: 
Terms of Props (e.g. proofs, theorem names) use `snake_case` &
When something named with `UpperCamelCase` is part of something named with snake_case, it is referenced in `lowerCamelCase`.

pfaffelh replied:
Thanks for pointing this out. I hope I have caught all occurrencies which disagreed with the convention.

JasonKYi commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -32,6 +33,9 @@ convergence in measure and other notions of convergence.
   which converges in measure to `g`, then `f` has a subsequence which convergence almost
   everywhere to `g`.
+* `MeasureTheory.exists_seq_tendstoInMeasure_atTop_iff`: for a sequence of functions `f`,
+convergence in measure is equivalent to the fact that every subsequence has another subsequence
+that converges almost surely.
```
```suggestion
  convergence in measure is equivalent to the fact that every subsequence has another subsequence
  that converges almost surely.
```

JasonKYi commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -219,27 +258,142 @@ theorem TendstoInMeasure.exists_seq_tendsto_ae (hfg : TendstoInMeasure μ f atTo
+
+end
+
+lemma forall_seq_tendstoInMeasure_atTop {u : Filter ι} {v : Filter κ} {f : ι → α → E}
```
I don't quite understand this name. I don't see an at top filter here. Maybe the name `TendstoInMeasure.comp_of_tendsto` is a better name?

JasonKYi commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -219,27 +258,142 @@ theorem TendstoInMeasure.exists_seq_tendsto_ae (hfg : TendstoInMeasure μ f atTo
+
+lemma forall_seq_tendstoInMeasure_atTop {u : Filter ι} {v : Filter κ} {f : ι → α → E}
+    {g : α → E} {ns : κ → ι} (hfg : TendstoInMeasure μ f u g) (hns : Tendsto ns v u) :
+    TendstoInMeasure μ (fun n => f (ns n)) v g :=
```
```suggestion
    TendstoInMeasure μ (f ∘ ns) v g :=
```

JasonKYi commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -219,27 +258,142 @@ theorem TendstoInMeasure.exists_seq_tendsto_ae (hfg : TendstoInMeasure μ f atTo
+variable {f : ℕ → α → E} {g h : α → E}
+
+-- The LimitInMeasure is ae unique
+theorem ae_unique_of_limitInMeasure' (hg : TendstoInMeasure μ f atTop g)
```
```suggestion
theorem TendstoInMeasure.ae_eq' (hg : TendstoInMeasure μ f atTop g)
```

JasonKYi commented on Mathlib/Topology/Instances/ENNReal.lean:
```diff
@@ -106,6 +111,21 @@ theorem tendsto_toNNReal {a : ℝ≥0∞} (ha : a ≠ ∞) :
 
+theorem tendsto_toNNReal_iff {ι : Type*} {f : ι → ℝ≥0∞} {u : Filter ι} {a : ℝ≥0∞}
+    (ha : a ≠ ∞) (hf : ∀ x, f x ≠ ∞) : Tendsto f u (𝓝 a) ↔
+    Tendsto (ENNReal.toNNReal ∘ f ) u (nhds (a.toNNReal)) := by
```
```suggestion
    Tendsto (ENNReal.toNNReal ∘ f ) u (𝓝 (a.toNNReal)) := by
```

EtienneC30 commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -58,6 +62,41 @@ theorem tendstoInMeasure_iff_norm [SeminormedAddCommGroup E] {l : Filter ι} {f
+    exact h ε hε
+
+-- Convergence in measure is stable under taking subsequences.
+lemma TendstoInMeasure.mono [Dist E] {f : ι → α → E}  {g : α → E} {u v : Filter ι} (huv : v ≤ u)
+    (hg :  TendstoInMeasure μ f u g) : TendstoInMeasure μ f v g :=
```
```suggestion
lemma TendstoInMeasure.mono [Dist E] {f : ι → α → E} {g : α → E} {u v : Filter ι} (huv : v ≤ u)
    (hg : TendstoInMeasure μ f u g) : TendstoInMeasure μ f v g :=
```

pfaffelh replied:
Done.

EtienneC30 commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -58,6 +62,41 @@ theorem tendstoInMeasure_iff_norm [SeminormedAddCommGroup E] {l : Filter ι} {f
+    (hg :  TendstoInMeasure μ f u g) : TendstoInMeasure μ f v g :=
+  fun ε hε => (hg ε hε).mono_left huv
+
+lemma TendstoInMeasure.subseq [Dist E] {f : ℕ → α → E}  {g : α → E} {ns : ℕ → ℕ}
+    (hns : StrictMono ns) (hg :  TendstoInMeasure μ f atTop g) :
```
```suggestion
lemma TendstoInMeasure.subseq [Dist E] {f : ℕ → α → E} {g : α → E} {ns : ℕ → ℕ}
    (hns : StrictMono ns) (hg : TendstoInMeasure μ f atTop g) :
```

pfaffelh replied:
Done.

EtienneC30 commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -58,6 +62,41 @@ theorem tendstoInMeasure_iff_norm [SeminormedAddCommGroup E] {l : Filter ι} {f
+  intro ε hε
+  apply Filter.Tendsto.comp (hg ε hε) (StrictMono.tendsto_atTop hns)
+
+lemma TendstoInMeasure.subseq' [Dist E] {f : ι → α → E}  {g : α → E} {u : Filter ι}
+    {v : Filter κ} {ns : κ → ι} (hns : Tendsto ns v u) (hg :  TendstoInMeasure μ f u g) :
```
```suggestion
lemma TendstoInMeasure.subseq' [Dist E] {f : ι → α → E} {g : α → E} {u : Filter ι}
    {v : Filter κ} {ns : κ → ι} (hns : Tendsto ns v u) (hg : TendstoInMeasure μ f u g) :
```

pfaffelh replied:
Done.

EtienneC30 commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -58,6 +62,41 @@ theorem tendstoInMeasure_iff_norm [SeminormedAddCommGroup E] {l : Filter ι} {f
+
+lemma TendstoInMeasure.subseq [Dist E] {f : ℕ → α → E}  {g : α → E} {ns : ℕ → ℕ}
+    (hns : StrictMono ns) (hg :  TendstoInMeasure μ f atTop g) :
+    TendstoInMeasure μ (f ∘ ns) atTop g := by
+  intro ε hε
+  apply Filter.Tendsto.comp (hg ε hε) (StrictMono.tendsto_atTop hns)
```
```suggestion
    TendstoInMeasure μ (f ∘ ns) atTop g :=
  fun ε hε => (hg ε hε).comp hns.tendsto_atTop
```

pfaffelh replied:
Done.

EtienneC30 commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -58,6 +62,41 @@ theorem tendstoInMeasure_iff_norm [SeminormedAddCommGroup E] {l : Filter ι} {f
+
+lemma TendstoInMeasure.subseq' [Dist E] {f : ι → α → E}  {g : α → E} {u : Filter ι}
+    {v : Filter κ} {ns : κ → ι} (hns : Tendsto ns v u) (hg :  TendstoInMeasure μ f u g) :
+    TendstoInMeasure μ (f ∘ ns) v g := by
+  intro ε hε
+  apply Filter.Tendsto.comp (hg ε hε) hns
```
```suggestion
    TendstoInMeasure μ (f ∘ ns) v g :=
  fun ε hε => (hg ε hε).comp hns
```

pfaffelh replied:
Done.

EtienneC30 commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -220,27 +259,142 @@ theorem TendstoInMeasure.exists_seq_tendsto_ae (hfg : TendstoInMeasure μ f atTo
 theorem TendstoInMeasure.exists_seq_tendstoInMeasure_atTop {u : Filter ι} [NeBot u]
     [IsCountablyGenerated u] {f : ι → α → E} {g : α → E} (hfg : TendstoInMeasure μ f u g) :
-    ∃ ns : ℕ → ι, TendstoInMeasure μ (fun n => f (ns n)) atTop g := by
+    ∃ (ns : ℕ → ι) (_ : Tendsto ns atTop u),  TendstoInMeasure μ (fun n => f (ns n)) atTop g := by
```
I think this is nicer.
```suggestion
    ∃ ns : ℕ → ι, Tendsto ns atTop u ∧ TendstoInMeasure μ (fun n => f (ns n)) atTop g := by
```

pfaffelh replied:
Done.

EtienneC30 commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -220,27 +259,142 @@ theorem TendstoInMeasure.exists_seq_tendsto_ae (hfg : TendstoInMeasure μ f atTo
-  obtain ⟨ms, hms⟩ := hfg.exists_seq_tendstoInMeasure_atTop
-  obtain ⟨ns, -, hns⟩ := hms.exists_seq_tendsto_ae
-  exact ⟨ms ∘ ns, hns⟩
+    ∃ (ns : ℕ → ι) (_ : Tendsto ns atTop u),
+    ∀ᵐ x ∂μ, Tendsto (fun i => f (ns i) x) atTop (𝓝 (g x)) := by
+  obtain ⟨ms, hms1, hms2⟩ := hfg.exists_seq_tendstoInMeasure_atTop
+  obtain ⟨ns, hns1, hns2⟩ := hms2.exists_seq_tendsto_ae
+  exact ⟨ms ∘ ns, Tendsto.comp hms1 (StrictMono.tendsto_atTop hns1), hns2⟩
```
This is nicer.
```suggestion
    ∃ ns : ℕ → ι, Tendsto ns atTop u ∧ ∀ᵐ x ∂μ, Tendsto (fun i => f (ns i) x) atTop (𝓝 (g x)) := by
  obtain ⟨ms, hms1, hms2⟩ := hfg.exists_seq_tendstoInMeasure_atTop
  obtain ⟨ns, hns1, hns2⟩ := hms2.exists_seq_tendsto_ae
  exact ⟨ms ∘ ns, hms1.comp hns1.tendsto_atTop, hns2⟩
```

pfaffelh replied:
Yes, it is. Done.

EtienneC30 commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -220,27 +259,142 @@ theorem TendstoInMeasure.exists_seq_tendsto_ae (hfg : TendstoInMeasure μ f atTo
+
+section
+
+/- An auxiliary lemma for a proof by contradiction in exists_seq_tendstoInMeasure_atTop_iff -/
```
```suggestion
/- An auxiliary lemma for a proof by contradiction in `exists_seq_tendstoInMeasure_atTop_iff`. -/
```

EtienneC30 commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -220,27 +259,142 @@ theorem TendstoInMeasure.exists_seq_tendsto_ae (hfg : TendstoInMeasure μ f atTo
+
+end
+
+lemma TendstoInMeasure.comp_of_tendsto {u : Filter ι} {v : Filter κ} {f : ι → α → E}
```
This is the same as `TendstoInMeasure.subseq'`.

pfaffelh replied:
True. Deleted.

EtienneC30 commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -220,27 +259,142 @@ theorem TendstoInMeasure.exists_seq_tendsto_ae (hfg : TendstoInMeasure μ f atTo
+    TendstoInMeasure μ (f ∘ ns) v g :=
+  fun ε hε => (hfg ε hε).comp hns
+
+lemma subseq_of_not_tendsto {f : ℕ → NNReal} (h : ¬Tendsto f atTop (𝓝 (0 : ℝ≥0))) :
```
This lemma is also too specific, the argument should be done directly in the proof.

EtienneC30 commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -220,27 +259,142 @@ theorem TendstoInMeasure.exists_seq_tendsto_ae (hfg : TendstoInMeasure μ f atTo
+
+/- TendstoInMeasure is equivalent to  every subsequence has another subsequence
+which converges almost surely. -/
+theorem exists_seq_tendstoInMeasure_atTop_iff (hfin : MeasureTheory.IsFiniteMeasure μ)
+    {f : ℕ → α → E} (hf : ∀ (n : ℕ), AEStronglyMeasurable (f n) μ) {g : α → E} :
+    (TendstoInMeasure μ f atTop g) ↔
+    ∀ (ns : ℕ → ℕ) (_ : StrictMono ns), ∃ (ns' : ℕ → ℕ) (_ : StrictMono ns'), ∀ᵐ (ω : α) ∂μ,
+    Tendsto (fun i ↦ f (ns (ns' i)) ω) atTop (𝓝 (g ω)) := by
```
Easier to read I think:
```suggestion
theorem exists_seq_tendstoInMeasure_atTop_iff [IsFiniteMeasure μ]
    {f : ℕ → α → E} (hf : ∀ (n : ℕ), AEStronglyMeasurable (f n) μ) {g : α → E} :
    (TendstoInMeasure μ f atTop g) ↔
      ∀ ns : ℕ → ℕ, StrictMono ns → ∃ ns' : ℕ → ℕ, StrictMono ns' ∧
        ∀ᵐ (ω : α) ∂μ, Tendsto (fun i ↦ f (ns (ns' i)) ω) atTop (𝓝 (g ω)) := by
```

pfaffelh replied:
Ok, done.

EtienneC30 commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -220,27 +259,142 @@ theorem TendstoInMeasure.exists_seq_tendsto_ae (hfg : TendstoInMeasure μ f atTo
+    Tendsto (fun i ↦ f (ns (ns' i)) ω) atTop (𝓝 (g ω)) := by
+  rw [tendstoInMeasure_iff_tendsto_toNNReal]
+  constructor
+  · intros hfg ns hns
+    have h1 : TendstoInMeasure μ (f ∘ ns)
+      atTop g := (TendstoInMeasure.subseq hns
+      (tendstoInMeasure_iff_tendsto_toNNReal.mpr hfg))
+    have ⟨ns', hns1, hns2⟩ :=
+    TendstoInMeasure.exists_seq_tendsto_ae h1
+    use ns', hns1
+    filter_upwards [hns2] with x hns3
+    exact hns3
```
This can be done in one line, provided that you move the above `rewrite below`:
```suggestion
  · exact fun hfg _ hns ↦ (hfg.subseq hns).exists_seq_tendsto_ae
```

EtienneC30 commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -220,27 +259,142 @@ theorem TendstoInMeasure.exists_seq_tendsto_ae (hfg : TendstoInMeasure μ f atTo
+    exact hns3
+  · rw [← not_imp_not]
+    intros h1
+    push_neg
+    obtain h2 : ∃ (ε : ℝ) (_ : 0 < ε),
+      ¬(Tendsto (fun n => (μ { x | ε ≤ dist (f n x) (g x) }).toNNReal) atTop (𝓝 0)) := by
+      · by_contra h3
+        apply h1
+        push_neg at h3
+        exact h3
+    rcases h2 with ⟨ε, hε, h4⟩
```
Here you can `push_neg` in `h1` too and automatically get `ε`.
```suggestion
    push_neg at *
    obtain ⟨ε, hε, h2⟩ := h1
```

EtienneC30 commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -220,27 +259,142 @@ theorem TendstoInMeasure.exists_seq_tendsto_ae (hfg : TendstoInMeasure μ f atTo
+        push_neg at h3
+        exact h3
+    rcases h2 with ⟨ε, hε, h4⟩
+    obtain h5 := @subseq_of_not_tendsto (fun n ↦ (μ {x | ε ≤ dist (f n x) (g x)}).toNNReal) h4
```
Here is a way to replace your lemma:
```suggestion
    obtain ⟨δ, ns, hδ, hns, h3⟩ : ∃ (δ : ℝ≥0) (ns : ℕ → ℕ), 0 < δ ∧ StrictMono ns ∧
        ∀ n, δ ≤ (μ {x | ε ≤ dist (f (ns n) x) (g x)}).toNNReal := by
      obtain ⟨s, hs, h4⟩ := not_tendsto_iff_exists_frequently_nmem.1 h2
      obtain ⟨δ, hδ, h5⟩ := NNReal.nhds_zero_basis.mem_iff.1 hs
      obtain ⟨ns, hns, h6⟩ := extraction_of_frequently_atTop h4
      exact ⟨δ, ns, hδ, hns, fun n ↦ Set.not_mem_Iio.1 (Set.not_mem_subset h5 (h6 n))⟩
```

EtienneC30 commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -220,27 +259,142 @@ theorem TendstoInMeasure.exists_seq_tendsto_ae (hfg : TendstoInMeasure μ f atTo
+    rcases h2 with ⟨ε, hε, h4⟩
+    obtain h5 := @subseq_of_not_tendsto (fun n ↦ (μ {x | ε ≤ dist (f n x) (g x)}).toNNReal) h4
+    rcases h5 with ⟨δ, hδ, ns, hns, h5⟩
+    use ns, hns
+    intros ns' _
```
You can do this in one line with `refine`. `refine` is very useful, it allows you to write the skeleton of the proof and creating a goal for each `?_` you write.
```suggestion
    refine ⟨ns, hns, fun ns' _ ↦ ?_⟩
```

EtienneC30 commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -220,27 +259,142 @@ theorem TendstoInMeasure.exists_seq_tendsto_ae (hfg : TendstoInMeasure μ f atTo
+    apply h4
+    have h4 : ∀ (n : ℕ), AEStronglyMeasurable ((f ∘ ns ∘ ns') n) μ := by exact fun n ↦
+      (hf ((ns ∘ ns') n))
+    have h8 := (MeasureTheory.tendstoInMeasure_of_tendsto_ae (f := f ∘ ns ∘ ns') (h4) h6)
```
h4 is unnecessary here.
```suggestion
    have h8 := tendstoInMeasure_of_tendsto_ae (fun _ ↦ hf _) h6
```

EtienneC30 commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -220,27 +259,142 @@ theorem TendstoInMeasure.exists_seq_tendsto_ae (hfg : TendstoInMeasure μ f atTo
+    have h4 : ∀ (n : ℕ), AEStronglyMeasurable ((f ∘ ns ∘ ns') n) μ := by exact fun n ↦
+      (hf ((ns ∘ ns') n))
+    have h8 := (MeasureTheory.tendstoInMeasure_of_tendsto_ae (f := f ∘ ns ∘ ns') (h4) h6)
+    obtain h7 := fun n => h5 (ns' n)
```
A bit shorter (`obtain` should only be used when you want to destruct the statement, typically if there is an `∃` or an `∧`) :
```suggestion
    have h7 n := h5 (ns' n)
```

EtienneC30 commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -220,27 +259,142 @@ theorem TendstoInMeasure.exists_seq_tendsto_ae (hfg : TendstoInMeasure μ f atTo
+    use ns, hns
+    intros ns' _
+    by_contra h6
+    apply h4
```
This is not needed.

EtienneC30 commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -220,27 +259,142 @@ theorem TendstoInMeasure.exists_seq_tendsto_ae (hfg : TendstoInMeasure μ f atTo
+    have h8 := (MeasureTheory.tendstoInMeasure_of_tendsto_ae (f := f ∘ ns ∘ ns') (h4) h6)
+    obtain h7 := fun n => h5 (ns' n)
+    rw [tendstoInMeasure_iff_tendsto_toNNReal] at h8
+    exfalso
+    revert h7
+    apply false_of_tendsto_of_bddBelow_aux
+      (fun n => (μ {x | ε ≤ dist (f (ns (ns' n)) x) (g x)}).toNNReal) δ hδ (h8 ε hε)
```
This finishes the proof:
```suggestion
    exact lt_irrefl _ (lt_of_le_of_lt (ge_of_tendsto' (h5 ε hε) h6) hδ)
```

EtienneC30 commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -220,27 +259,142 @@ theorem TendstoInMeasure.exists_seq_tendsto_ae (hfg : TendstoInMeasure μ f atTo
+    revert h7
+    apply false_of_tendsto_of_bddBelow_aux
+      (fun n => (μ {x | ε ≤ dist (f (ns (ns' n)) x) (g x)}).toNNReal) δ hδ (h8 ε hε)
+
+section
+
+end
```
You can remove that.

EtienneC30 commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -220,27 +259,142 @@ theorem TendstoInMeasure.exists_seq_tendsto_ae (hfg : TendstoInMeasure μ f atTo
+variable {f : ℕ → α → E} {g h : α → E}
+
+-- The TendstoInMeasure is ae unique
+theorem ae_unique_of_tendstoInMeasure' (hg : TendstoInMeasure μ f atTop g)
```
It is unnecessary to have this lemma when the general version is easily deduced, just put the proof in the general case directly.

EtienneC30 commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -220,27 +259,142 @@ theorem TendstoInMeasure.exists_seq_tendsto_ae (hfg : TendstoInMeasure μ f atTo
+theorem ae_unique_of_tendstoInMeasure {g h : α → E} {f : ι → α → E}  {u : Filter ι} [NeBot u]
+    [IsCountablyGenerated u] (hg : TendstoInMeasure μ f u g) (hh : TendstoInMeasure μ f u h) :
+    g =ᵐ[μ] h := by
+  obtain ⟨ns,h1,h1'⟩ := TendstoInMeasure.exists_seq_tendstoInMeasure_atTop hg
```
```suggestion
  obtain ⟨ns, h1, h1'⟩ := hg.exists_seq_tendstoInMeasure_atTop
```

EtienneC30 commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -220,27 +259,142 @@ theorem TendstoInMeasure.exists_seq_tendsto_ae (hfg : TendstoInMeasure μ f atTo
+    [IsCountablyGenerated u] (hg : TendstoInMeasure μ f u g) (hh : TendstoInMeasure μ f u h) :
+    g =ᵐ[μ] h := by
+  obtain ⟨ns,h1,h1'⟩ := TendstoInMeasure.exists_seq_tendstoInMeasure_atTop hg
+  exact ae_unique_of_tendstoInMeasure' (f := f ∘ ns) h1' (TendstoInMeasure.subseq' h1 hh)
```
```suggestion
  exact ae_unique_of_tendstoInMeasure' h1' (hh.subseq' h1)
```

EtienneC30 commented on Mathlib/Topology/Instances/ENNReal/Lemmas.lean:
```diff
@@ -51,6 +51,11 @@ theorem coe_range_mem_nhds : range ((↑) : ℝ≥0 → ℝ≥0∞) ∈ 𝓝 (r
 theorem continuous_coe : Continuous ((↑) : ℝ≥0 → ℝ≥0∞) :=
   isEmbedding_coe.continuous
 
+lemma tendsto_coe_id {a : ℝ≥0∞} (ha : a ≠ ⊤) : Tendsto (fun (x : ℝ≥0) => (x : ℝ≥0∞))
+    (nhds a.toNNReal) (nhds a) := by
+  nth_rewrite 2 [← coe_toNNReal ha]
+  exact ContinuousAt.tendsto (Continuous.continuousAt continuous_coe)
```
A better name maybe. This does not involve the identity so I don't think `id` should appear.
```suggestion
lemma tendsto_coe_toNNReal {a : ℝ≥0∞} (ha : a ≠ ⊤) : Tendsto (↑) (𝓝 a.toNNReal) (𝓝 a) := by
  nth_rewrite 2 [← coe_toNNReal ha]
  exact continuous_coe.tendsto _
```

EtienneC30 commented on Mathlib/Topology/Instances/ENNReal/Lemmas.lean:
```diff
@@ -79,6 +84,21 @@ theorem tendsto_toNNReal {a : ℝ≥0∞} (ha : a ≠ ∞) :
 
+theorem tendsto_toNNReal_iff {ι : Type*} {f : ι → ℝ≥0∞} {u : Filter ι} {a : ℝ≥0∞}
+    (ha : a ≠ ∞) (hf : ∀ x, f x ≠ ∞) : Tendsto f u (𝓝 a) ↔
+    Tendsto (ENNReal.toNNReal ∘ f ) u (𝓝 (a.toNNReal)) := by
```
```suggestion
    Tendsto (ENNReal.toNNReal ∘ f) u (𝓝 (a.toNNReal)) := by
```

EtienneC30 commented on Mathlib/Topology/Instances/ENNReal/Lemmas.lean:
```diff
@@ -79,6 +84,21 @@ theorem tendsto_toNNReal {a : ℝ≥0∞} (ha : a ≠ ∞) :
+theorem tendsto_toNNReal_iff {ι : Type*} {f : ι → ℝ≥0∞} {u : Filter ι} {a : ℝ≥0∞}
+    (ha : a ≠ ∞) (hf : ∀ x, f x ≠ ∞) : Tendsto f u (𝓝 a) ↔
+    Tendsto (ENNReal.toNNReal ∘ f ) u (𝓝 (a.toNNReal)) := by
+  constructor
+  · exact fun h =>  Filter.Tendsto.comp (ENNReal.tendsto_toNNReal ha) h
+  · intro h
```
`refine` is great! Do not hesitate to use dot notation as frequently as possible. If a theorem starts with `Filter.Tendsto` and takes an argument `Filter.Tendsto _ _ _` you can replace `Filter.Tendsto` with your argument in the name of the lemma, making it shorter.
```suggestion
  refine ⟨fun h => (ENNReal.tendsto_toNNReal ha).comp h, fun h => ?_⟩
```

EtienneC30 commented on Mathlib/Topology/Instances/ENNReal/Lemmas.lean:
```diff
@@ -79,6 +84,21 @@ theorem tendsto_toNNReal {a : ℝ≥0∞} (ha : a ≠ ∞) :
+    exact h2
+
+theorem tendsto_toNNReal_iff' {ι : Type*} {f : ι → ℝ≥0∞} {u : Filter ι} {a : ℝ≥0}
+    (hf : ∀ x, f x ≠ ∞): Tendsto f u (𝓝 a) ↔ Tendsto (ENNReal.toNNReal ∘ f ) u (𝓝 a) := by
```
```suggestion
    (hf : ∀ x, f x ≠ ∞) : Tendsto f u (𝓝 a) ↔ Tendsto (ENNReal.toNNReal ∘ f) u (𝓝 a) := by
```

EtienneC30 commented on Mathlib/Topology/MetricSpace/Pseudo/Defs.lean:
```diff
@@ -832,6 +832,15 @@ theorem tendsto_atTop' [Nonempty β] [SemilatticeSup β] [NoMaxOrder β] {u : β
 
+/-- This variant can be useful for proofs by contradiction.
+-/
+lemma false_of_Tendsto_of_bddBelow  [Nonempty β] [SemilatticeSup β] {u : β → α}
```
I am not sure this lemma is very useful, although I would like someone else's opinion.

EtienneC30 reviewed (commented):
Here comes a second wave! More little details about the shape of the code. Some lemmas are unnecessary I think, I put advices to shorten many proofs. In particular the proof of `exists_seq_tendstoInMeasure_atTop_iff` can be made much shorter. I tried to explain in some places, but please do not hesitate to ask if something is not clear, e.g. why one way would be better than another, or if you do not understand the syntax of some tactics.

pfaffelh commented on Mathlib/MeasureTheory/Function/AEEqFun.lean:
```diff
@@ -863,6 +863,17 @@ theorem coeFn_posPart (f : α →ₘ[μ] γ) : ⇑(posPart f) =ᵐ[μ] fun a =>
+section AELimit
+
+-- The ae-limit is ae-unique.
+theorem tendsto_ae_unique {α ι E : Type*} [TopologicalSpace E] [T2Space E]
+    {x : MeasurableSpace α} {μ : Measure α} {g h : α → E} {f : ι → α → E} {l : Filter ι} [l.NeBot]
+    (hg : ∀ᵐ ω ∂μ, Tendsto (fun i => f i ω) l (𝓝 (g ω)))
+    (hh : ∀ᵐ ω ∂μ, Tendsto (fun i => f i ω) l (𝓝 (h ω))) : g =ᵐ[μ] h := by
```
Yes, this is much better.

pfaffelh commented on Mathlib/Data/ENNReal/Basic.lean:
```diff
@@ -201,6 +201,12 @@ theorem coe_toNNReal : ∀ {a : ℝ≥0∞}, a ≠ ∞ → ↑a.toNNReal = a
   | ⊤, h => (h rfl).elim
 
+@[simp]
+theorem coe_of_fun_toNNReal {ι : Type*} {f : ι → ℝ≥0∞} (hf : ∀ x, f x ≠ ∞) :
```
Ok, done.

pfaffelh commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -58,6 +62,41 @@ theorem tendstoInMeasure_iff_norm [SeminormedAddCommGroup E] {l : Filter ι} {f
       ∀ ε, 0 < ε → Tendsto (fun i => μ { x | ε ≤ ‖f i x - g x‖ }) l (𝓝 0) := by
   simp_rw [TendstoInMeasure, dist_eq_norm]
 
+theorem tendstoInMeasure_iff_tendsto_toNNReal [Dist E] {_ : MeasurableSpace α} {μ : Measure α}
+    [hfin: MeasureTheory.IsFiniteMeasure μ] {f : ι → α → E} {l : Filter ι} {g : α → E} :
+    TendstoInMeasure μ f l g ↔ ∀ ε, 0 < ε → Tendsto (fun i =>
+    (μ { x | ε ≤ dist (f i x) (g x) }).toNNReal) l (𝓝 0) := by
```
Thanks, this is shorter indeed.

pfaffelh commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -58,6 +62,41 @@ theorem tendstoInMeasure_iff_norm [SeminormedAddCommGroup E] {l : Filter ι} {f
+  have hfin : ∀ ε, ∀ i, μ { x | ε ≤ dist (f i x) (g x) } ≠ ⊤ := by
+    exact fun ε i ↦ (measure_ne_top μ {x | ε ≤ dist (f i x) (g x)})
+  constructor
+  · intros h ε hε
```
I streamlined the code by using `refine ⟨fun h ε hε ↦ ?_, fun h ε hε ↦ ?_⟩`.

pfaffelh commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -58,6 +62,41 @@ theorem tendstoInMeasure_iff_norm [SeminormedAddCommGroup E] {l : Filter ι} {f
+    exact fun ε i ↦ (measure_ne_top μ {x | ε ≤ dist (f i x) (g x)})
+  constructor
+  · intros h ε hε
+    have hf : (fun i =>
+    (μ { x | ε ≤ dist (f i x) (g x) }).toNNReal) = ENNReal.toNNReal ∘ (fun i =>
+    (μ { x | ε ≤ dist (f i x) (g x) })) := by
+      exact rfl
```
Done.

pfaffelh commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -58,6 +62,41 @@ theorem tendstoInMeasure_iff_norm [SeminormedAddCommGroup E] {l : Filter ι} {f
+    rw [hf, ← ENNReal.tendsto_toNNReal_iff' (hfin ε)]
+    exact h ε hε
+  · intros h ε hε
+    rw [ENNReal.tendsto_toNNReal_iff ENNReal.zero_ne_top (hfin ε) ]
```
Done.

pfaffelh commented:
Thanks for all the comments, which helped me to learn several things (in particular streamlining some proofs).

EtienneC30 commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -220,27 +252,84 @@ theorem TendstoInMeasure.exists_seq_tendsto_ae (hfg : TendstoInMeasure μ f atTo
+        ∀ᵐ (ω : α) ∂μ, Tendsto (fun i ↦ f (ns (ns' i)) ω) atTop (𝓝 (g ω)) := by
+  refine ⟨fun hfg _ hns ↦ (hfg.subseq hns).exists_seq_tendsto_ae, not_imp_not.mp (fun h1 ↦ ?_)⟩
+  rw [tendstoInMeasure_iff_tendsto_toNNReal] at h1
+  simp only [not_forall, Classical.not_imp, not_exists, not_and,
+    not_eventually] at *
```
You can just `push_neg` here.
```suggestion
  push_neg at *
```

EtienneC30 commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -220,27 +252,84 @@ theorem TendstoInMeasure.exists_seq_tendsto_ae (hfg : TendstoInMeasure μ f atTo
+    obtain ⟨ns, hns, h6⟩ := extraction_of_frequently_atTop h4
+    exact ⟨δ, ns, hδ, hns, fun n ↦ Set.not_mem_Iio.1 (Set.not_mem_subset h5 (h6 n))⟩
+  refine ⟨ns, hns, fun ns' _ ↦ ?_⟩
+  by_contra h6
+  simp only [not_frequently, not_not] at h6
```
You can combine those with `by_contra!`, which use `by_contra` and automatically applies `push_neg` at the new hypothesis.
```suggestion
  by_contra! h6
```

EtienneC30 commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -220,27 +252,84 @@ theorem TendstoInMeasure.exists_seq_tendsto_ae (hfg : TendstoInMeasure μ f atTo
+  refine ⟨ns, hns, fun ns' _ ↦ ?_⟩
+  by_contra h6
+  simp only [not_frequently, not_not] at h6
+  obtain h7 := tendstoInMeasure_iff_tendsto_toNNReal.mp <|
+    MeasureTheory.tendstoInMeasure_of_tendsto_ae (f := f ∘ ns ∘ ns') (fun n ↦
+    (hf ((ns ∘ ns') n))) h6
```
Nicer I think:
```suggestion
  obtain h7 := tendstoInMeasure_iff_tendsto_toNNReal.mp <|
    tendstoInMeasure_of_tendsto_ae (fun n ↦ hf _) h6
```

EtienneC30 commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -220,27 +252,84 @@ theorem TendstoInMeasure.exists_seq_tendsto_ae (hfg : TendstoInMeasure μ f atTo
+  obtain h7 := tendstoInMeasure_iff_tendsto_toNNReal.mp <|
+    MeasureTheory.tendstoInMeasure_of_tendsto_ae (f := f ∘ ns ∘ ns') (fun n ↦
+    (hf ((ns ∘ ns') n))) h6
+  have h (f : ℕ → ℝ≥0) (hf1 : Tendsto f atTop (𝓝 0)) : ¬(∀ (n : ℕ), δ ≤ f n) := by
+    simp only [imp_false, not_forall, not_le]
+    obtain ⟨x, h2⟩ : ∃ a, ∀ (b : ℕ), a ≤ b → dist (f b) 0 < ↑δ := by
+      simp_rw [← Metric.mem_ball, ← Set.mem_preimage, ← mem_atTop_sets, ← mem_map]
+      exact hf1 (Metric.ball_mem_nhds 0 hδ)
+    use x
+    obtain h3 := h2 x (by rfl)
+    simp only [dist, NNReal.val_eq_coe, NNReal.coe_zero, sub_zero, NNReal.abs_eq,
+      NNReal.coe_lt_coe] at h3
+    exact h3
+  exact h (fun n => (μ {x | ε ≤ dist (f (ns (ns' n)) x) (g x)}).toNNReal)
+      (h7 ε hε) <| fun m ↦ h3 (ns' m)
```
You can just end the proof with this.
```suggestion
  exact lt_irrefl _ (lt_of_le_of_lt (ge_of_tendsto' (h7 ε hε) (fun n ↦ h3 _)) hδ)
```

EtienneC30 commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -220,27 +252,84 @@ theorem TendstoInMeasure.exists_seq_tendsto_ae (hfg : TendstoInMeasure μ f atTo
+variable [MetricSpace E]
+variable {f : ℕ → α → E} {g h : α → E}
+
+/-- The TendstoInMeasure is ae unique -/
```
```suggestion
/-- The limit in measure is ae unique. -/
```

EtienneC30 replied:
You forgot the full stop at the end.

EtienneC30 commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -220,27 +252,84 @@ theorem TendstoInMeasure.exists_seq_tendsto_ae (hfg : TendstoInMeasure μ f atTo
+variable {f : ℕ → α → E} {g h : α → E}
+
+/-- The TendstoInMeasure is ae unique -/
+theorem ae_unique_of_tendstoInMeasure {g h : α → E} {f : ι → α → E} {u : Filter ι} [NeBot u]
```
Following the convention:
```suggestion
theorem tendstoInMeasure_ae_unique {g h : α → E} {f : ι → α → E} {u : Filter ι} [NeBot u]
```

EtienneC30 commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -220,27 +252,84 @@ theorem TendstoInMeasure.exists_seq_tendsto_ae (hfg : TendstoInMeasure μ f atTo
+    [IsCountablyGenerated u] (hg : TendstoInMeasure μ f u g) (hh : TendstoInMeasure μ f u h) :
+    g =ᵐ[μ] h := by
+  obtain ⟨ns, h1, h1'⟩ := hg.exists_seq_tendsto_ae'
+  obtain ⟨ns', h2, h2'⟩ :=
+    (TendstoInMeasure.subseq' h1 hh).exists_seq_tendsto_ae'
```
```suggestion
  obtain ⟨ns', h2, h2'⟩ := (TendstoInMeasure.subseq' h1 hh).exists_seq_tendsto_ae'
```

EtienneC30 commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -220,27 +252,84 @@ theorem TendstoInMeasure.exists_seq_tendsto_ae (hfg : TendstoInMeasure μ f atTo
+  obtain ⟨ns, h1, h1'⟩ := hg.exists_seq_tendsto_ae'
+  obtain ⟨ns', h2, h2'⟩ :=
+    (TendstoInMeasure.subseq' h1 hh).exists_seq_tendsto_ae'
+  obtain h4 : ∀ᵐ (x : α) ∂μ, Tendsto (fun i ↦ (f ∘ ns) (ns' i) x) atTop (𝓝 (g x)) := by
+    filter_upwards [h1'] with ω h'
+    apply h'.comp h2
+  filter_upwards [h4, h2'] with ω hg1 hh1
+  exact tendsto_nhds_unique hg1 hh1
```
You can do this directly:
```suggestion
  filter_upwards [h1', h2'] with ω hg1 hh1
  exact tendsto_nhds_unique (hg1.comp h2) hh1
```
When writing a proof, it is normal to write it in details, and having many `have` blocks to structure your argument, it makes it easier to write I think. Once you're done with the proof, do not hesitate to go through it and try to identify parts which can be shortened, `have` which can be gotten rid of by inlining their proofs etc. It is not easy to do at first, it comes with practice, but it makes the proof easier to read.

EtienneC30 commented on Mathlib/Topology/Instances/ENNReal/Lemmas.lean:
```diff
@@ -50,6 +50,11 @@ theorem coe_range_mem_nhds : range ((↑) : ℝ≥0 → ℝ≥0∞) ∈ 𝓝 (r
 theorem continuous_coe : Continuous ((↑) : ℝ≥0 → ℝ≥0∞) :=
   isEmbedding_coe.continuous
 
+lemma tendsto_coe_toNNReal {a : ℝ≥0∞} (ha : a ≠ ⊤) : Tendsto (fun (x : ℝ≥0) => (x : ℝ≥0∞))
+    (nhds a.toNNReal) (nhds a) := by
```
You can write it like this:
```suggestion
lemma tendsto_coe_toNNReal {a : ℝ≥0∞} (ha : a ≠ ⊤) : Tendsto (↑) (𝓝 a.toNNReal) (𝓝 a) := by
```
The `↑` notation is the coercion map. Here you can use it because the types are inferred thanks to the filters. Also please make sure to use the notation for the neighbourhood filter when possible.

EtienneC30 replied:
You did not remove the line skip.

EtienneC30 commented on Mathlib/Topology/Instances/ENNReal/Lemmas.lean:
```diff
@@ -50,6 +50,11 @@ theorem coe_range_mem_nhds : range ((↑) : ℝ≥0 → ℝ≥0∞) ∈ 𝓝 (r
+lemma tendsto_coe_toNNReal {a : ℝ≥0∞} (ha : a ≠ ⊤) : Tendsto (fun (x : ℝ≥0) => (x : ℝ≥0∞))
+    (nhds a.toNNReal) (nhds a) := by
+  nth_rewrite 2 [← coe_toNNReal ha]
+  exact ContinuousAt.tendsto (Continuous.continuousAt continuous_coe)
```
Shorter
```suggestion
  exact continuous_coe.tendsto _
```

EtienneC30 reviewed (commented):
More remarks on the style, more shortcuts. I noticed that there are some stuff I had indicated last time which you did not take into account. When you decide not to apply a suggestion, can you please indicate it, and potentially explain why? (readability etc.) Some of them are merely suggestions and up to your preference, but I don't know if you missed them or if you just decided not to apply them. I am aware that there are a lot of remarks and it takes time to go through it, but if you don't settle all the suggestions before removing the awaiting author label it will only slow down he process even more.
Also note that you have a build error, please make sure to fix it before removing awaiting author.

pfaffelh commented:
Sorry that I apparently have not addressed all issues from the last round, since I did my best to do so. Usually, I respond, but nevertheless click on "resolve", since the page becomes too long otherwise. Is this ok?

EtienneC30 commented:
If you apply a suggestion and then resolve, it's fine, and you don't need to respond to indicate that you applied it because I can see it in the code. On the other hand, there are some suggestions which you only applied partially, and I don't know if it is because you missed them or because you preferred not to apply them. So I would say that the best way to proceed is to make sure you applied the whole suggestion before resolving, and if you had a good reason not too, then indicate why and keep it unresolved so I can see it easily.
Of course there are many comments and it is easy to oversee some, especially when there are many changes in one suggestion.
If a code suggestion seems reasonable to you, you can use the "Commit suggestion" button to automatically commit the change to the branch (make sure to refresh your local branch if you do that). This avoids forgetting some changes.

EtienneC30 commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -220,27 +252,84 @@ theorem TendstoInMeasure.exists_seq_tendsto_ae (hfg : TendstoInMeasure μ f atTo
+  obtain ⟨ns, hns1, hns2⟩ := hms2.exists_seq_tendsto_ae
+  exact ⟨ms ∘ ns, hms1.comp hns1.tendsto_atTop, hns2⟩
+
+/-- TendstoInMeasure is equivalent to  every subsequence has another subsequence
```
For example here you changed the "has" but you did not remove the whitespace and add the "\`" around `TendstoInMeasure`.

EtienneC30 replied:
You did not add the backticks around `TendstoInMeasure`. They allow the name to be displayed code-style in the doc.

EtienneC30 commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -58,6 +62,34 @@ theorem tendstoInMeasure_iff_norm [SeminormedAddCommGroup E] {l : Filter ι} {f
+
+lemma TendstoInMeasure.subseq' [Dist E] {f : ι → α → E} {g : α → E} {u : Filter ι}
+    {v : Filter κ} {ns : κ → ι} (hns : Tendsto ns v u) (hg : TendstoInMeasure μ f u g) :
+    TendstoInMeasure μ (f ∘ ns) v g := fun ε hε ↦ (hg ε hε).comp hns
```
You didn't do the renaming.

EtienneC30 commented:
Please fix the build before removing the awaiting-author label.

pfaffelh commented:
Sorry, now it builds... (Forgot to save my file...)

EtienneC30 commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -58,6 +62,29 @@ theorem tendstoInMeasure_iff_norm [SeminormedAddCommGroup E] {l : Filter ι} {f
+  · rw [ENNReal.tendsto_toNNReal_iff ENNReal.zero_ne_top (hfin ε)]
+    exact h ε hε
+
+/-- Convergence in measure is stable under taking subsequences. -/
+lemma TendstoInMeasure.mono [Dist E] {f : ι → α → E} {g : α → E} {u v : Filter ι} (huv : v ≤ u)
```
Can you rephrase or remove the docstring here please? This is not really about subsequences.

EtienneC30 commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -220,27 +247,67 @@ theorem TendstoInMeasure.exists_seq_tendsto_ae (hfg : TendstoInMeasure μ f atTo
+￼which converges almost surely. -/
+theorem exists_seq_tendstoInMeasure_atTop_iff (hfin : MeasureTheory.IsFiniteMeasure μ)
+    {f : ℕ → α → E} (hf : ∀ (n : ℕ), AEStronglyMeasurable (f n) μ) {g : α → E} :
+    (TendstoInMeasure μ f atTop g) ↔
```
You can drop the parentheses here.
```suggestion
    TendstoInMeasure μ f atTop g ↔
```

EtienneC30 commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -220,27 +247,67 @@ theorem TendstoInMeasure.exists_seq_tendsto_ae (hfg : TendstoInMeasure μ f atTo
+    exact ⟨δ, ns, hδ, hns, fun n ↦ Set.not_mem_Iio.1 (Set.not_mem_subset h5 (h6 n))⟩
+  refine ⟨ns, hns, fun ns' _ ↦ ?_⟩
+  by_contra h6
+  obtain h7 := tendstoInMeasure_iff_tendsto_toNNReal.mp <|
```
`obtain` should only be used to destruct a statement, like an "exists" or an "and" or an "or". Otherwise use `have`.
```suggestion
  have h7 := tendstoInMeasure_iff_tendsto_toNNReal.mp <|
```

EtienneC30 commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -220,27 +247,67 @@ theorem TendstoInMeasure.exists_seq_tendsto_ae (hfg : TendstoInMeasure μ f atTo
+section TendstoInMeasureUnique
+
+variable [MetricSpace E]
+variable {f : ℕ → α → E} {g h : α → E}
```
The variables here are only available in the section, i.e. in the theorem below. But you reintroduced them in the statement. Either remove the `variable` line, or remove the variables in the statement.

EtienneC30 commented on Mathlib/Topology/Instances/ENNReal/Lemmas.lean:
```diff
@@ -78,6 +82,18 @@ theorem tendsto_toNNReal {a : ℝ≥0∞} (ha : a ≠ ∞) :
   rw [nhds_coe, tendsto_map'_iff]
   exact tendsto_id
 
+theorem tendsto_toNNReal_iff {ι : Type*} {f : ι → ℝ≥0∞} {u : Filter ι} {a : ℝ≥0∞}
+    (ha : a ≠ ∞) (hf : ∀ x, f x ≠ ∞) : Tendsto f u (𝓝 a) ↔
+    Tendsto (ENNReal.toNNReal ∘ f) u (𝓝 (a.toNNReal)) := by
```
You can reuse the variables which are already available in the file. In this case you can use the general type `α` and a variable `a` is already available. Also, we prefer the "simpler" form on the right of the iff. This also matches the name of the lemma best. Here's a header modified accordingly:
```suggestion
theorem tendsto_toNNReal_iff {f : α → ℝ≥0∞} {u : Filter α} (ha : a ≠ ∞) (hf : ∀ x, f x ≠ ∞) :
    Tendsto (ENNReal.toNNReal ∘ f) u (𝓝 (a.toNNReal)) ↔ Tendsto f u (𝓝 a) := by
```
The same can be applied to the lemma below. In VSCode, if you want to know which variables are available anywhere in the file, you can type `#where` at the place you are interested in.

EtienneC30 commented on Mathlib/Topology/Instances/ENNReal/Lemmas.lean:
```diff
@@ -78,6 +82,18 @@ theorem tendsto_toNNReal {a : ℝ≥0∞} (ha : a ≠ ∞) :
+theorem tendsto_toNNReal_iff {ι : Type*} {f : ι → ℝ≥0∞} {u : Filter ι} {a : ℝ≥0∞}
+    (ha : a ≠ ∞) (hf : ∀ x, f x ≠ ∞) : Tendsto f u (𝓝 a) ↔
+    Tendsto (ENNReal.toNNReal ∘ f) u (𝓝 (a.toNNReal)) := by
+  refine ⟨fun h =>  Filter.Tendsto.comp (ENNReal.tendsto_toNNReal ha) h, fun h => ?_ ⟩
```
Here you can use dot notation. There are also unnecessary spaces.
```suggestion
  refine ⟨fun h => (ENNReal.tendsto_toNNReal ha).comp h, fun h => ?_⟩
```

EtienneC30 replied:
You did not apply this. Do you prefer to avoid dot notation? In general I think it improves readability. Also there is an unnecessary white space after `=>`.

EtienneC30 commented on Mathlib/Topology/Instances/ENNReal/Lemmas.lean:
```diff
@@ -78,6 +82,18 @@ theorem tendsto_toNNReal {a : ℝ≥0∞} (ha : a ≠ ∞) :
+  rw [← coe_comp_toNNReal_comp hf]
+  exact (tendsto_coe_toNNReal ha).comp h
+
+theorem tendsto_toNNReal_iff' {ι : Type*} {f : ι → ℝ≥0∞} {u : Filter ι} {a : ℝ≥0}
+    (hf : ∀ x, f x ≠ ∞) : Tendsto f u (𝓝 a) ↔ Tendsto (ENNReal.toNNReal ∘ f) u (𝓝 a) := by
```
Same remarks as previously.

EtienneC30 reviewed (commented):
A few more comments, we're almost there!

EtienneC30 commented on Mathlib/Topology/Instances/ENNReal/Lemmas.lean:
```diff
@@ -78,6 +82,17 @@ theorem tendsto_toNNReal {a : ℝ≥0∞} (ha : a ≠ ∞) :
+  rw [← coe_comp_toNNReal_comp hf]
+  exact (tendsto_coe_toNNReal ha).comp h
+
+theorem tendsto_toNNReal_iff' {f : α → ℝ≥0∞} {u : Filter α} {a : ℝ≥0}
+    (hf : ∀ x, f x ≠ ∞) : Tendsto (ENNReal.toNNReal ∘ f) u (𝓝 a) ↔ Tendsto f u (𝓝 a) := by
+  rw [← @toNNReal_coe a]
```
Here is a nicer header. In general try to put the statement on its own line if possible. Also the `@` is not needed here.
```suggestion
theorem tendsto_toNNReal_iff' {f : α → ℝ≥0∞} {u : Filter α} {a : ℝ≥0} (hf : ∀ x, f x ≠ ∞) :
    Tendsto (ENNReal.toNNReal ∘ f) u (𝓝 a) ↔ Tendsto f u (𝓝 a) := by
  rw [← toNNReal_coe a]
```

EtienneC30 commented:
Thanks a lot, that's great!
maintainer merge

pfaffelh commented:
Thanks, EtienneC30!

YaelDillies commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -220,27 +246,64 @@ theorem TendstoInMeasure.exists_seq_tendsto_ae (hfg : TendstoInMeasure μ f atTo
+
+/-- `TendstoInMeasure` is equivalent to every subsequence having another subsequence
+￼which converges almost surely. -/
+theorem exists_seq_tendstoInMeasure_atTop_iff (hfin : MeasureTheory.IsFiniteMeasure μ)
```
```suggestion
theorem exists_seq_tendstoInMeasure_atTop_iff (hfin : IsFiniteMeasure μ)
```
and please `open MeasureTheory` if necessary

YaelDillies commented on Mathlib/Topology/Instances/ENNReal/Lemmas.lean:
```diff
@@ -50,6 +50,10 @@ theorem coe_range_mem_nhds : range ((↑) : ℝ≥0 → ℝ≥0∞) ∈ 𝓝 (r
   isEmbedding_coe.continuous
 
+lemma tendsto_coe_toNNReal {a : ℝ≥0∞} (ha : a ≠ ⊤) : Tendsto (↑) (𝓝 a.toNNReal) (𝓝 a) := by
+  nth_rewrite 2 [← coe_toNNReal ha]
+  exact continuous_coe.tendsto _
```
A nice style for this kind of proofs where things match up up to some simplifications is
```suggestion
  simpa [coe_toNNReal ha] using continuous_coe.tendsto _
```
or in this case possibly even
```suggestion
  simpa [ha] using continuous_coe.tendsto _
```

YaelDillies commented on Mathlib/MeasureTheory/Function/ConvergenceInMeasure.lean:
```diff
@@ -58,6 +62,28 @@ theorem tendstoInMeasure_iff_norm [SeminormedAddCommGroup E] {l : Filter ι} {f
+theorem tendstoInMeasure_iff_tendsto_toNNReal [Dist E] [IsFiniteMeasure μ]
+    {f : ι → α → E} {l : Filter ι} {g : α → E} :
+    TendstoInMeasure μ f l g ↔
+    ∀ ε, 0 < ε → Tendsto (fun i => (μ { x | ε ≤ dist (f i x) (g x) }).toNNReal) l (𝓝 0) := by
```
```suggestion
      ∀ ε, 0 < ε → Tendsto (fun i => (μ { x | ε ≤ dist (f i x) (g x) }).toNNReal) l (𝓝 0) := by
```

YaelDillies reviewed (commented):
Sorry, got a few last minute requests!

RemyDegenne commented:
Thanks! I was going to merge but I'll let you decide if you want to use Yaël's proof suggestion and then merge yourself.
bors d+

pfaffelh commented:
bors r+
